core/payments/ondemand/on_demand_payment_validator.go

```diff
@@ -0,0 +1,179 @@
+		logger:            logger,
+		ledgers:           cache,
+		onChainState:      onChainState,
+		dynamoClient:      dynamoClient,
+		onDemandTableName: onDemandTableName,
+	}, nil
+}
+
+// Debit validates an on-demand payment for a blob dispersal
+// The caller is responsible for verifying the signature before calling this method
+func (pv *OnDemandPaymentValidator) Debit(
+	ctx context.Context,
+	accountID gethcommon.Address,
+	symbolCount uint32,
+	quorumNumbers []uint8,
+) error {
+	ledger, err := pv.getOrCreateLedger(ctx, accountID)
+	if err != nil {
+		return fmt.Errorf("get or create on-demand ledger: %w", err)
+	}
+
+	_, err = ledger.Debit(ctx, symbolCount, quorumNumbers)
+	if err != nil {
+		return fmt.Errorf("debit on-demand payment: %w", err)
+	}
+
+	return nil
+}
+
+// getOrCreateLedger gets an existing on-demand ledger from the cache, or creates a new one if it doesn't exist
+func (pv *OnDemandPaymentValidator) getOrCreateLedger(
+	ctx context.Context,
+	accountID gethcommon.Address,
+) (*OnDemandLedger, error) {
+	// Fast path: check if ledger already exists in cache
+	if ledger, exists := pv.ledgers.Get(accountID); exists {
+		return ledger, nil
+	}
+
+	// Slow path: acquire lock and check again
+	pv.ledgerCreationLock.Lock()
+	defer pv.ledgerCreationLock.Unlock()
+
+	if ledger, exists := pv.ledgers.Get(accountID); exists {
+		return ledger, nil
+	}
+
+	onDemandPayment, err := pv.onChainState.GetOnDemandPaymentByAccount(ctx, accountID)
+	if err != nil {
+		return nil, fmt.Errorf("get on-demand payment for account %v: %w", accountID.Hex(), err)
+	}
+
+	pricePerSymbol := big.NewInt(int64(pv.onChainState.GetPricePerSymbol()))
+	minNumSymbols := pv.onChainState.GetMinNumSymbols()
+
+	cumulativePaymentStore, err := NewCumulativePaymentStore(pv.dynamoClient, pv.onDemandTableName, accountID)
+	if err != nil {
+		return nil, fmt.Errorf("new cumulative payment store: %w", err)
+	}
+
+	newLedger, err := OnDemandLedgerFromStore(
+		ctx,
+		onDemandPayment.CumulativePayment,
+		pricePerSymbol,
+		minNumSymbols,
+		cumulativePaymentStore,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create on-demand ledger: %w", err)
+	}
+
+	pv.ledgers.Add(accountID, newLedger)
+	return newLedger, nil
+}
+
+// UpdateTotalDeposits updates the total deposits for multiple accounts.
+//
+// Will attempt to make all updates, even if one update fails. A multierror is returned, describing any/all errors
+// that occurred during the updates.
+func (pv *OnDemandPaymentValidator) UpdateTotalDeposits(updates []TotalDepositUpdate) error {
+	if len(updates) == 0 {
+		return nil
+	}
+
+	var result *multierror.Error
+	for _, update := range updates {
+		ledger, exists := pv.ledgers.Get(update.AccountAddress)
+		if !exists {
+			// if we aren't already tracking the account, there's nothing to do. we'll start tracking it if the
+			// account ever makes an on-demand dispersal
+			continue
+		}
+
+		err := ledger.UpdateTotalDeposits(update.NewTotalDeposit)
```

Comment:
each of these is making a network call to write to dynamodb correct?
If that's the case then consider using an https://pkg.go.dev/golang.org/x/sync/errgroup to make the requests in parallel instead.